### PR TITLE
Add support for `scatter_add` gradient

### PR DESCRIPTION
**Pull request by rdspring1**
This PR adds support for `scatter_add` gradient. 

**Note:** The value gradient is only defined when `value.shape == index.shape`.
See https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.scatter_add_

### CODE REVIEW

File: thunder/tests/opinfos.py [hunk 1]
```diff
@@ -4679,13 +4679,17 @@ def scatter_add_sample_generator(op, device, dtype, requires_grad, **kwargs):
     # torch.scatter_add expects index to be long but not int
     # Index is not differentiable! Marking requires_grad as False
     make_index = partial(make_tensor, device=device, dtype=torch.long, requires_grad=False)
-    # Not sure if we need to consider higher order gradient, marking requires_grad as False
-    make_source = partial(make_tensor, device=device, dtype=dtype, requires_grad=False)
+    make_source = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
 
+    # NOTE The value gradient is only valid when src.shape == index.shape.
```
Comment (mruberry) [inline thunder/tests/opinfos.py:4684]: This is a really nice comment for future readers

File: thunder/tests/opinfos.py [hunk 2]
```diff
@@ -4705,13 +4709,38 @@ def scatter_add_sample_generator(op, device, dtype, requires_grad, **kwargs):
...
+    shape_b = (3, 2, 3)
+    shape_source = (4, 3, 9)
+
+    a = make(shape_a)
+    b = make_index(shape_b, low=0, high=shape_a[dim])
+    c = make_source(shape_source)
+    yield (
+        SampleInput(a, index=b, src=c, dim=dim),
+        RuntimeError,
+        "The gradient for the src Tensor is implemented only when src.shape == index.shape. src shape is (.*?) while index shape is (.*?).",
```
Comment (mruberry) [inline thunder/tests/opinfos.py:4735]: Is the gradient well-defined when the source and index have different shapes? If so, why do we only support this case?

File: thunder/core/transforms.py [hunk 3]
```diff
@@ -853,6 +853,21 @@ def _gather_prim_grad(a: TensorProxy, index: TensorProxy, dim: int) -> TensorPro
 register_grad(pids.GATHER, _gather_prim_grad)
 
 
+@torchctx
+def _scatter_add_prim_grad(a: TensorProxy, /, index: TensorProxy, value: TensorProxy, dim: int) -> TensorProxy:
+    fwd = prims.scatter_add(a, index, value, dim)
+
+    g = get_grad(fwd)
+    # NOTE The value gradient is only valid when src.shape == index.shape.
```
Comment (mruberry) [inline thunder/core/transforms.py:861]: (which was checked in forward)

**Review by mruberry — APPROVED** [review]
Cool!

@IvanYashchuk and @nikitaved , would you like to take a look, too?

File: thunder/core/transforms.py (hunk 3 above)
Comment (nikitaved) [reply thunder/core/transforms.py:861]: No, forward does not have this lim. This is the "hidden" requirement to make sure the grad is correct. It is an issue with a long history :)

File: thunder/tests/opinfos.py (hunk 2 above)
Comment (nikitaved) [reply thunder/tests/opinfos.py:4735]: It is defined, but not with this implementation. See https://github.com/pytorch/pytorch/issues/27614#issuecomment-564648819, for example.

File: thunder/tests/opinfos.py (hunk 1 above)
Comment (nikitaved) [reply thunder/tests/opinfos.py:4684]: super minor nit: I would probably change "valid" to "correct"

File: thunder/torch/__init__.py [hunk 4]
```diff
@@ -2100,6 +2100,11 @@ def gather(a: TensorLike, /, dim: int, index: TensorLike) -> TensorLike:
 # NOTE PyTorch's scatter_add has a parameter named 'src', not 'source'
 @torchsymbol(torch.scatter_add, is_method=True)
 def scatter_add(a: TensorLike, /, dim: int, index: TensorLike, src: TensorLike) -> TensorLike:
+    utils.check(
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply thunder/torch/__init__.py:2103]: It is a limitation of the backward implementation - so I would suggest putting this check there.

[the remark this answers is not in the document]
Comment (rdspring1) [reply thunder/torch/__init__.py]: I moved the check to the grad transform.

**Comment by mruberry** [comment]
@t-vi @Borda Looks like this will require overriding the CI to merge
